### PR TITLE
Post merge health checks

### PR DESCRIPTION
**Pull request by deffrian**
Fixes | Closes | Resolves #

> Anything marked as optional that you didn't need to fill in, please remove it from the PR description. Choose one of the keywords above to refer to the issue this PR solves, followed by the issue number (e.g Fixes # 666). If there is no issue, remove the line. Remove this note after reading.

## Changes:
- When Merge is enabled Health Checks plugin verifies that CL is alive

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [ x ] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ x ] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs [hunk 1]
```diff
@@ -69,30 +80,50 @@ public CheckHealthResult CheckHealth()
             long netPeerCount = _syncServer.GetPeerCount();
             SyncingResult syncingResult = _ethSyncingInfo.GetFullInfo();
 
-            if (_isMining == false && syncingResult.IsSyncing)
+            if (_mergeConfig.Enabled)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:83]: the better could be not null in TTD

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs [hunk 2]
```diff
@@ -69,30 +80,50 @@ public CheckHealthResult CheckHealth()
...
+                    AddStillSyncingMessage(messages, syncingResult);
+                }
+                else
+                {
+                    AddFullySyncMessage(messages);
+                }
                 CheckPeers(messages, netPeerCount);
+
+                healthy = !syncingResult.IsSyncing & CheckClAlive(messages,
+                    _blockFinder.Head!.Difficulty == UInt256.Zero
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:96]: Nice, I think we should check if CL is alive also when node is synced

Comment (MarekM25) [reply src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:96]: Maybe we should add newPayload postMerge

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs [hunk 3]
```diff
@@ -110,6 +141,37 @@ public CheckHealthResult CheckHealth()
                    _healthHintService.MaxSecondsIntervalForProducingBlocksHint();
         }
 
+        private bool _previousClCheckResult = true;
+        private DateTime _previousClCheckTime = DateTime.MinValue;
+        private int _lastMethodCallSuccesses;
+
+        private bool CheckClAlive(ICollection<(string Description, string LongDescription)> messages, string methodName)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:148]: I thought that we will abstract this check and if post-merge we will run the node without CL we will print some warnings to user, something with meaning: "Dear user! Don't forget to run your CL" :)

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs [hunk 4]
```diff
@@ -110,6 +141,37 @@ public CheckHealthResult CheckHealth()
...
+        {
+            var now = _api.Timestamper.UtcNow;
+            var forkChoiceUpdatedStats = _api.JsonRpcLocalStats!.GetMethodStats(methodName);
+
+            if (forkChoiceUpdatedStats.Successes == _lastMethodCallSuccesses)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:153]: to say that CL is alive it doesn't have to be the success, but to say that our node is healthy it has to

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs [hunk 5]
```diff
@@ -69,30 +80,50 @@ public CheckHealthResult CheckHealth()
...
+                }
+                else
+                {
+                    AddFullySyncMessage(messages);
+                }
                 CheckPeers(messages, netPeerCount);
+
+                healthy = !syncingResult.IsSyncing & CheckClAlive(messages,
+                    _blockFinder.Head!.Difficulty == UInt256.Zero
+                        ? "engine_forkchoiceUpdatedV1"
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:97]: just add newPayload and exchangeTransition here?

File: src/Nethermind/Nethermind.HealthChecks/HealthChecksConfig.cs [hunk 6]
```diff
@@ -27,9 +27,11 @@ public class HealthChecksConfig : IHealthChecksConfig
...
         public string WebhooksRestorePayload { get; set; } = "{\"attachments\":[{\"color\":\"#36a64f\",\"pretext\":\"Health Check Status :+1:\",\"fields\":[{\"title\":\"Details\",\"value\":\"`More details available at /healthchecks-ui`\",\"short\":false},{\"title\":\"description\",\"value\":\"The HealthCheck `[[LIVENESS]]` is recovered. All is up and running\",\"short\":false}]}]}";
         public bool UIEnabled { get; set; } = false;
-        
+
         public ulong? MaxIntervalWithoutProcessedBlock { get; set; }
-        
+
         public ulong? MaxIntervalWithoutProducedBlock { get; set; }
+
+        public int MaxIntervalClRequestTime { get; set; } = 20;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/HealthChecksConfig.cs:35]: This interval should be longer, at least 5-8 times slot time

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs (hunk 4 above)
Comment (deffrian) [reply src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs]: Why do we need that?

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthChecks.cs [hunk 7]
```diff
@@ -20,41 +20,51 @@
...
             _nodeHealthService = nodeHealthService ?? throw new ArgumentNullException(nameof(nodeHealthService));
+            _api = api;
             _logger = logManager.GetClassLogger();
         }
 
         public Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
         {
             try
             {
+                if (_api.SpecProvider!.TerminalTotalDifficulty != null && !_nodeHealthService.CheckClAlive())
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthChecks.cs:48]: I thought that we will add this log outside the healthchecks plugin, but I think it is fine for now :)

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs [hunk 8]
```diff
@@ -69,30 +77,54 @@ public CheckHealthResult CheckHealth()
...
-                bool peers = CheckPeers(messages, netPeerCount);
-                bool processing = IsProcessingBlocks(messages);
-                bool producing = IsProducingBlocks(messages);
-                healthy = peers && processing && producing;
+                if (!_isMining && syncingResult.IsSyncing)
+                {
+                    AddStillSyncingMessage(messages, syncingResult);
+                    CheckPeers(messages, netPeerCount);
+                }
+                else if (!_isMining && !syncingResult.IsSyncing)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:108]: why we're not checking if CL is alive if the node is not syncing?

**Review by MarekM25 — CHANGES_REQUESTED** [review]
pls check if CL is alive when it is not syncing

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs [hunk 9]
```diff
@@ -110,6 +142,50 @@ public CheckHealthResult CheckHealth()
                    _healthHintService.MaxSecondsIntervalForProducingBlocksHint();
         }
 
+        public bool CheckClAlive()
+        {
+            if (_blockFinder.Head!.Difficulty == UInt256.Zero)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:147]: this logic is wrong. Imagine your node is syncing. Your Head = 0, but you will receive fcUs and newPayloads. If you have finalTotalDifficulty (PosSwitcher.FinalTotalDifficulty) (so the merge happened you can check for fcU and newPayloads only. In other cases please check all three methods: fcu, newPayload, exchangeTransition

**Review by MarekM25 — CHANGES_REQUESTED** [review]
Please read comments

File: src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs (hunk 8 above)
Comment (deffrian) [reply src/Nethermind/Nethermind.HealthChecks/NodeHealthService.cs:108]: It's legacy code, if merge is not enabled

File: src/Nethermind/Nethermind.HealthChecks/HealthChecksConfig.cs [hunk 10]
```diff
@@ -27,9 +27,11 @@ public class HealthChecksConfig : IHealthChecksConfig
...
         public string WebhooksRestorePayload { get; set; } = "{\"attachments\":[{\"color\":\"#36a64f\",\"pretext\":\"Health Check Status :+1:\",\"fields\":[{\"title\":\"Details\",\"value\":\"`More details available at /healthchecks-ui`\",\"short\":false},{\"title\":\"description\",\"value\":\"The HealthCheck `[[LIVENESS]]` is recovered. All is up and running\",\"short\":false}]}]}";
         public bool UIEnabled { get; set; } = false;
-        
+
         public ulong? MaxIntervalWithoutProcessedBlock { get; set; }
-        
+
         public ulong? MaxIntervalWithoutProducedBlock { get; set; }
+
+        public int MaxIntervalClRequestTime { get; set; } = 70;
```
Comment (kamilchodola) [inline src/Nethermind/Nethermind.HealthChecks/HealthChecksConfig.cs:35]: Need to add docs for this parameter.

File: src/Nethermind/Nethermind.HealthChecks/IHealthChecksConfig.cs [hunk 11]
```diff
@@ -44,11 +44,14 @@ public interface IHealthChecksConfig : IConfig
...
-        
+
         [ConfigItem(Description = "Max interval in seconds in which we assume that node processing blocks in a healthy way", DefaultValue = "null")]
         public ulong? MaxIntervalWithoutProcessedBlock { get; set; }
-        
+
         [ConfigItem(Description = "Max interval in seconds in which we assume that node producing blocks in a healthy way", DefaultValue = "null")]
         public ulong? MaxIntervalWithoutProducedBlock { get; set; }
+
+        [ConfigItem(Description = "Max request interval in witch we assume that CL works in a healthy way", DefaultValue = "70")]
```
Comment (kamilchodola) [inline src/Nethermind/Nethermind.HealthChecks/IHealthChecksConfig.cs:54]: Maybe fix typo from "witch" to "which" :D

**Review by kamilchodola — APPROVED** [review]

**Review by MarekM25 — APPROVED** [review]

**Pull request merged**